added smappee component

hmn opened the pull request:
## Description:

Added components for https://www.smappee.com/

## Example entry for `configuration.yaml` (if applicable):
```yaml
smappee:
    host: 10.0.0.5
    client_id: YOUR_CLIENT_ID
    client_secret: YOUR_CLIENT_SECRET
    username: YOUR_MYSMAPPEE_USERNAME
    password: YOUR_MYSMAPPEE_PASSWORD
```

## Checklist:
  - [x] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

Commit:
added smappee component

ciotlosm committed:
Fixed pylint errors and a few use cases when starting up with invalid credentials

Added coverage omit

hmn committed:
Merge pull request #1 from ciotlosm/smappee

Fixed pylint errors and a few use cases when starting up with invalid…

ciotlosm committed:
Added support to run only locally
Added a few more sensors
Added more error handling

Better parsing and debug message

hmn committed:
Merge pull request #2 from ciotlosm/smappee

Added support to run only locally

Commit:
fixed smappee switch after local/remote support was added

Commit:
Merge branch 'dev' into smappee

* dev:
  upgrade schiene to 0.20 (#11504)
  Upgrade pywebpush to 1.5.0 (#11497)
  Fix time functions would throw errors in python scripts (#11414)
  Clean up Alexa.intent and DialogFlow.intent (#11492)
  Add missing configuration variables (#11390)
  Add new iGlo component (#11171)
  Upgrade pysnmp to 4.4.4 (#11485)

Commit:
Merge branch 'dev' into smappee

* dev:
  Address missed review comments for Dark Sky weather (#11520)
  Fix canary flaky test (#11519)
  Lazy loading of service descriptions (#11479)
  Add Dark Sky weather component (#11435)
  Snips (new) added speech response, parse snips/duration (#11513)
  More tolerant KNX component if gateway cant be connected (#11511)
  timer: include the remaining time in the state attributes (#11510)

Commit:
Merge branch 'dev' into smappee

* dev:
  Reconnect before mochad switch send command (#11296)
  New features for Owntracks device_tracker (#11480)

ciotlosm committed:
Smappee - update switches for local support (#3)

* Merged with local version

* Updated smappy library with the patched one

Fixed lint, added merge missing param

Fixed missing run for requirements_all.txt

Fixed lint

* Fixed on/off based on library. Reverted change used for testing stacktrace

* Fixed switches to work with both remote and local active

Fixed lint

Fixed switches

Fixed lint

Commit:
nothing to update per switch as the states are not saved by smappee system

Commit:
added better error handling for communication errors with smappee

Commit:
fixed lint errors

gieljnssns commented on homeassistant/components/smappee.py:
```diff
@@ -0,0 +1,290 @@
+"""
+Support for Digital Ocean.
```
Shouldn't  this be Smappee

Commit:
fixed comment

Commit:
fixed lint error

Commit:
Merge branch 'dev' into smappee

* dev: (85 commits)
  Bump dev to 0.62.0.dev0 (#11652)
  Add Deconz support for Zigbee green power devices like Hue Tap (#11455)
  Update Pyarlo to 0.1.2 (#11626)
  Use kelvin/mireds correctly for setting iglo white (#11622)
  patch stop command (#11612)
  Upgrade yarl to 0.18.0 (#11609)
  Xiaomi lib upgrade (#11603)
  Bugfix and cleanup for Rfxtrx (#11600)
  Concord232 alarm arm away fix (#11597)
  Snips add say and say_actions services (new) (#11596)
  Fix state for trigger with forced updates (#11595)
  Pushbullet email support (fix) (#11590)
  Fix Tahoma stop command for 2 types of shutters (#11588)
  Core support for hass.io calls & Bugfix check_config (#11571)
  Avoid returning empty media_image_url string (#11557)
  Upgrade coinmarketcap to 4.1.2 (#11634)
  Update CODEOWNERS
  Add templates to MQTT climate (#11623)
  upgrade xiaomi lib (#11629)
  Fixes and enhancements for the Tahoma platform (#11547)
  ...

dcsg commented:
+1

Commit:
Merge branch 'dev' into smappee

* dev: (313 commits)
  check_config check bootstrap errors (#12291)
  have climate fallback to state if no ATTR_OPERATION_MODE (#12271) (#12279)
  Update limitlessled requirement to v1.0.9 (#12275)
  Retry keyset cloud (#12270)
  allow wildcards in subscription (#12247)
  Update limitlessled requirement to v1.0.9 (#12275)
  Retry keyset cloud (#12270)
  Fix some rfxtrx devices with multiple sensors (#12264)
  allow wildcards in subscription (#12247)
  device_tracker.asuswrt: Clean up unused connection param (#12262)
  Version bump to 0.64.0.dev0
  Version bump to 0.63
  One bug fix and one improvement to the statistics sensor. (#12259)
  Update frontend to 20180209.0
  Fix duplicate entity_ids in System Monitor (#12124)
  Fix for iTunes media player not updating artwork (#12089)
  Update owntracks.py (#12260)
  Update pyhomematic to 0.1.39 (#12265)
  miflora - fix for exception handling bug (#12149)
  Fix source code using Windows newline (#12248)
  ...

Commit:
Merge branch 'dev' into smappee

* dev: (31 commits)
  Upgrade Sphinx to 1.7.0 (#12335)
  nmap_tracker: don't scan on setup (#12322)
  Fix platform dependencies (#12330)
  device_tracker.asuswrt: Ignore unreachable ip neigh entries (#12201)
  Unifi tracking filter by SSID (#12281)
  Always return lockState == LOCKED when handling Alexa.LockController (#12328)
  Fix line endings [skipci] (#12333)
  Fix Report State for Alexa Brightness Controller (#12318)
  Respect entity namespace for entity registry (#12313)
  zha: Add remove service (#11683)
  Revert #12316 (#12329)
  Updated RainMachine to play better with the entity registry (#12315)
  Remove unique ID from netatmo (#12317)
  Force LF line endings for Windows (#12266)
  Removed default sensor configuration (#12252)
  zha: Update zigpy-xbee to 0.0.2
  Protect bloomsky platform setup (#12316)
  Purge recorder data by default (#12271)
  add friendly_name_template to template sensor (#12268)
  Clarify tahoma errrors (#12307)
  ...

Commit:
Merge branch 'dev' into smappee

* dev: (40 commits)
  File Path fixes for RPi Camera (#12338)
  Downgrade limitlessled to 1.0.8 (#12403)
  Downgrade limitlessled to 1.0.8 (#12403)
  Introduce zone_id to identify player+zone (#12382)
  Updated beacontools (#12368)
  Fix WUnderground names (#12346)
  Fix MercedesMe - add check for unsupported features (#12342)
  Version bump to 0.63.2
  Changed pyvera version to 0.2.41 (#12391)
  Pollen.com: Entity Registry updates and cleanup (#12361)
  Improve service by allowing to reference entity id instead of deconz id (#11862)
  python-miio version bumped. (Closes: #12389, Closes: #12298) (#12392)
  Add attributes (fixes #12332) (#12377)
  Remove usage of deprecated assert method (#12379)
  Upgrade youtube_dl to 2018.02.11 (#12383)
  Introduce zone_id to identify player+zone (#12382)
  Specify algorithms for webpush jwt verification (#12378)
  SMA sensor add SSL and upgrade to pysma 0.2 (#12354)
  Upgrade panasonic_viera to 0.3.1 (#12370)
  Allow disabling entities in the registry (#12360)
  ...

hmn committed:
Merge branch 'dev' into smappee

* dev:
  Print every changed file on new line (#12412)
  Fix MQTT payload decode returning prematurely (#12420)
  Fixed 3 small issues in isy994 component (#12421)
  Extract data validator to own file and add tests (#12401)
  Upgrade sphinx-autodoc-typehints to 1.2.5 (#12404)
  Code cleanup of velux scene (#12390)

hmn committed:
fixed lint error

balloob commented on homeassistant/components/sensor/smappee.py:
```diff
@@ -0,0 +1,177 @@
+        self.data = None
+        self._state = None
+        self._timestamp = None
+        self.update()
```
Don't call `update` in constructor. Instead add `True` as second param to `add_devices`

balloob commented on homeassistant/components/sensor/smappee.py:
```diff
@@ -0,0 +1,177 @@
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return SENSOR_TYPES[self._sensor][1]
```
Can you cache this in the constructor so you don't have to do the same lookup on each update

balloob commented on homeassistant/components/sensor/smappee.py:
```diff
@@ -0,0 +1,177 @@
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return SENSOR_TYPES[self._sensor][3]
```
Same

balloob commented on homeassistant/components/sensor/smappee.py:
```diff
@@ -0,0 +1,177 @@
+        if self._location_id:
+            attr['Location Id'] = self._location_id
+            attr['Location Name'] = self._smappee.locations[self._location_id]
+        attr['Last Update'] = self._timestamp
```
Don't include this. States already have a last update attribute

balloob commented on homeassistant/components/sensor/smappee.py:
```diff
@@ -0,0 +1,177 @@
+        attr['Last Update'] = self._timestamp
+        return attr
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
```
Instead of this, just define SCAN_INTERVAL and this method will never be called more than once during that interval.

balloob commented on homeassistant/components/sensor/smappee.py:
```diff
@@ -0,0 +1,177 @@
+    def update(self):
+        """Get the latest data from Smappee and update the state."""
+        if self._sensor == 'alwaysOn':
+            data = self._smappee.get_consumption(
```
This is doing a call to the API per sensor per update (every 30 seconds!). Any way  you can fetch all data at once?

balloob commented on homeassistant/components/sensor/smappee.py:
```diff
@@ -0,0 +1,177 @@
+                self._state = round(power, 2)
+                self._timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+        else:
+            return None
```
Remove this. Does nothing.

balloob commented on homeassistant/components/smappee.py:
```diff
@@ -0,0 +1,290 @@
+        """Get switches from local Smappee."""
+        if self.is_local_active:
+            return self._localsmappy.load_command_control_config()
+        return False
```
Inconsistent return type

balloob commented on homeassistant/components/smappee.py:
```diff
@@ -0,0 +1,290 @@
+                _LOGGER.error(
+                    "Error getting comsumption from Smappee cloud. (%s)",
+                    error)
+                return False
```
Inconsistent return type

balloob commented on homeassistant/components/sensor/smappee.py:
```diff
@@ -0,0 +1,177 @@
+                self._timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+                self._state = round(consumption.get('solar') / 1000, 2)
+        elif self._sensor == 'power_today':
+            data = self._smappee.get_consumption(
```
There is no error handling being done here. This method can return False.

balloob commented on homeassistant/components/smappee.py:
```diff
@@ -0,0 +1,290 @@
+        # 3 = daily values,
+        # 4 = monthly values,
+        # 5 = quarterly values
+        if self.is_remote_active:
```
Use guard clause

```python
if not self.is_remote_active:
    return

end = …
```

balloob commented on homeassistant/components/smappee.py:
```diff
@@ -0,0 +1,290 @@
+        # 3 = daily values,
+        # 4 = monthly values,
+        # 5 = quarterly values
+        if self.is_remote_active:
```
guard clause

balloob commented on homeassistant/components/smappee.py:
```diff
@@ -0,0 +1,290 @@
+                    error)
+                return False
+
+        return False
```
These error values are never handled. Probably better to propagate the error to the update method of the sensor.

balloob commented on homeassistant/components/smappee.py:
```diff
@@ -0,0 +1,290 @@
+            self._smappy.actuator_on(location_id, actuator_id, duration)
+        else:
+            self._localsmappy.on_command_control(actuator_id)
+            return True
```
Inconsistent return

balloob commented on homeassistant/components/smappee.py:
```diff
@@ -0,0 +1,290 @@
+
+    def instantaneous_values(self):
+        """ReportInstantaneousValues."""
+        if self.is_local_active:
```
guard clause

balloob commented on homeassistant/components/switch/smappee.py:
```diff
@@ -0,0 +1,99 @@
+    # Skip local switches if remote available
+    if not smappee.is_remote_active:
+        local_devices = smappee.local_devices
+        if smappee.is_local_active:
```
Merge with parent if statement.

balloob commented on homeassistant/components/switch/smappee.py:
```diff
@@ -0,0 +1,99 @@
+
+    def turn_on(self, **kwargs):
+        """Turn on Comport Plug."""
+        self._smappee.actuator_on(self._location_id, self._switch_id,
```
Why do this twice?

balloob commented on homeassistant/components/switch/smappee.py:
```diff
@@ -0,0 +1,99 @@
+                                  self._remoteswitch)
+        self._smappee.actuator_on(self._location_id, self._switch_id,
+                                  self._remoteswitch)
+        self._state = True
```
Why are you setting state to True? You can only do that if you check the command result and make sure the command was successful.

Also, Home Assistant calls your update method after calling a service, which will also query your device

hmn replied:
The plugs don't have any way of communicating with the rest of the smappee system, they work by just sending a signal out and then hoping it is received.

It's not really very nice so setting the state seemed to be the only way I could handle them, if theres a better way to handle switches where you can't get the status just point me in that direction and I'll look into it

balloob commented:
Please include a link to the docs PR.

hmn commented:
Docs PR is here https://github.com/home-assistant/home-assistant.github.io/pull/4354

hmn committed:
update smappee module with reviewer comments
- update smappy module
- cache cloud api requests
- added actuator info
- updated return states

hmn commented:
Thanks for al of the feedback, I tried fixing all of it, but I'm not exactly a python expert, so I hope it's better :)

balloob commented:
Awesome work! 🎉
